Python/tier2_redundancy_eliminator_bytecodes.c

```diff
@@ -189,6 +177,58 @@ dummy_func(void) {
         }
     }
 
+    op(_TO_BOOL, (value -- res)) {
+        (void)value;
+        OUT_OF_SPACE_IF_NULL(res = sym_new_known_type(ctx, &PyBool_Type));
+    }
+
+    op(_TO_BOOL_BOOL, (value -- value)) {
+        if (sym_matches_type(value, &PyBool_Type)) {
+            REPLACE_OP(this_instr, _NOP, 0, 0);
+        }
+        else {
+            sym_set_type(value, &PyBool_Type);
+        }
+    }
+
+    op(_TO_BOOL_INT, (value -- res)) {
+        sym_set_type(value, &PyLong_Type);
+        if (is_const(value)) {
+            PyObject *load = _PyLong_IsZero((PyLongObject *)get_const(value))
+                             ? Py_False : Py_True;
+            REPLACE_OP(this_instr, _POP_TOP_LOAD_CONST_INLINE_BORROW, 0, (uintptr_t)load);
+            OUT_OF_SPACE_IF_NULL(res = sym_new_const(ctx, load));
+        }
+        else {
+            OUT_OF_SPACE_IF_NULL(res = sym_new_known_type(ctx, &PyBool_Type));
+        }
+    }
+
+    op(_TO_BOOL_LIST, (value -- res)) {
+        sym_set_type(value, &PyList_Type);
+        OUT_OF_SPACE_IF_NULL(res = sym_new_known_type(ctx, &PyBool_Type));
+    }
+
+    op(_TO_BOOL_NONE, (value -- res)) {
+        if (get_const(value) == Py_None) {
+            REPLACE_OP(this_instr, _POP_TOP_LOAD_CONST_INLINE_BORROW, 0, (uintptr_t)Py_None);
+        }
+        sym_set_type(value, &_PyNone_Type);
+        OUT_OF_SPACE_IF_NULL(res = sym_new_const(ctx, Py_False));
+    }
+
+    op(_TO_BOOL_STR, (value -- res)) {
+        sym_set_type(value, &PyUnicode_Type);
```

Comment:
It just occurred to me that `sym_set_type()` could check that the symbol isn't already known to have a different type that's inconsistent -- that would be either an `assert()` (if we know this can never happen unless there's a bug in the optimizer or trace generator) or it could just abort the optimization process and let the machinery fall back to Tier 1 (if we feel there are legitimate cases that could lead to such an inconsistency).

Reply:
See above comment